CC2650STK Debugger Support

nicholas-c-jackson opened the pull request:
Added additional support for the XDS100v3 debugger platform

Implemented default flash configuration for the XDS110 with support for XDS100v3 debugger through DEBUGGER make flags.

Will update CC2650STK wiki on closed pull request.

nicholas-c-jackson commented:
I would also like to add that I have tested the XDS100v3 with the CC2650STK and the LAUNCHXL-CC2650 and confirmed to be working.

PeterKietzmann commented:
@yogo1212, @fvcoen could you test this, do yo have the debugger platform available?

yogo1212 commented:
@PeterKietzmann 
don't have the hardware :-(
if you're willing to wait for a few days before merging i could buy it, though

PeterKietzmann commented:
@yogo1212 thanks for your responsiveness! Sure I could wait some days but there is really no need for you to buy it, only for testing :-)

fvcoen commented:
@PeterKietzmann I have access to the [SmartRF06 Evaluation Board](http://www.ti.com/tool/smartrf06ebk) which features the XDS100v3. I would be more than happy to help :) However I'm very inexperienced with debug probe so I would need some pointers to set things up...

nicholas-c-jackson commented:
@fvcoen I'm able to help, just need some details, what OS are you running?

Yonezawa-T2 commented:
I also don't have XDS100v3. Can anyone else be the assignee?

PeterKietzmann commented:
@fvcoen great :-)! I guess @nicholas-c-jackson can help you. @Yonezawa-T2 no one of "us" has it but if you feel uncomfortable with care-taking this PR you can reassign, as always

fvcoen commented:
@nicholas-c-jackson I'm running Ubuntu 14.04

nicholas-c-jackson commented:
@fvcoen Ok, I don't know what stage your environment is at, so I will start from the beginning.

- First, ensure you have TI's uniflash installed.
- Add the uniflash binaries to the UNIFLASH_PATH environment variable (which is the root folder for uniflash, containing uniflash.sh)
- Make example (hello world) `make BOARD=cc2650stk`

Which is all the steps here [https://github.com/RIOT-OS/RIOT/wiki/Board%3A-CC2650STK](https://github.com/RIOT-OS/RIOT/wiki/Board%3A-CC2650STK)

Where it differentiates is the following.

There is the default DEBUGGER make flag which will use the XDS110 which can be used by either;

`make BOARD=cc2650stk flash`

or 

`make BOARD=cc2650stk flash DEBUGGER=XDS110`

Which is case sensitive for the make flag.

To use the XDS100v3 run the same but with the other debugger make flag as so;

`make BOARD=cc2650stk flash DEBUGGER=XDS100v3`

Assuming your drivers are all setup, it should flash fine.

fvcoen commented:
@nicholas-c-jackson Sorry for the delay, I've been quite busy for the past few days.
I had no problem flashing the board with `DEBUGGER=XDS110` and `DEBUGGER=XDS100v3` :) I'm currently working on configuring gdb to actually use the debugger...

nicholas-c-jackson commented:
@fvcoen No problem, I have been following the progress you and Leon have been making on the RF core. I too have been super busy this week. I'm hoping soon I can jump on board with the development of the RF core. I've pushed a new commit changing the $DEBUGGER flag to $XDEBUGGER as there was a conflict with the `make debug` make script. 

Unfamiliarity on my part.

I have yet to use GDB with either of the debuggers. I will work on that tomorrow my time (GMT+10) and get back to you.

A-Paul commented:
I might also need some advise, but I have a *CC2650STK* here. I'll try to follow https://github.com/RIOT-OS/RIOT/pull/5419#issuecomment-219260570 tomorrow.

A-Paul commented:
> I also don't have XDS100v3. Can anyone else be the assignee?

@Yonezawa-T2, I relieve you of that burden. :-)

nicholas-c-jackson commented:
@A-Paul I have recently updated the flag to use XDEBUGGER, updated my original comment as well.

Yonezawa-T2 commented:
@A-Paul Thank you 😄 .

nicholas-c-jackson commented:
@A-Paul So I have spent some time researching about openocd and the cc26xx platform. I have seen that they have added support for that class of cpu's but I have yet to get them to work with the XDS debuggers. I have posted some questions on a few forums such as the TI and openocd forums, hopefully I will receive some additional information shortly.

I will look into additional means of JTAG based debugging for the cc26xx platform, although having continuity would be potentially more beneficial with riot and the openocd platform. Hopefully soon I can push the openocd configuration for the cc26xx and the XDS debuggers.

A-Paul commented:
@nicholas-c-jackson have you also noticed this: https://github.com/JelmerT/cc2538-bsl#cc26xx-and-cc13xx? Although it's about flashing and not debugging, it seems that TI was somewhat creative in using the *cc2560* on the *Sensortag*.

nicholas-c-jackson commented:
@A-Paul I looked into it. Shame that it doesn't work with sensor tag and I assume the launch xl variant.

nicholas-c-jackson commented:
For the mean time, I have paused trying to get openocd to work. I have managed to get gdb working with the emulator agents provided in uniflash for both the cc2650 sensor tag and the launch xl with the XDS100v3 and the XDS110. I will spend the next couple of days testing that everything is working correctly and package it into the unified make system. Hopefully this will assist @yogo1212 and @Yonezawa-T2 with the RFC that I plan to get involved with soon.

Nicholas

nicholas-c-jackson commented:
@A-Paul I have added the last of the commits for this pull request. 

The process of debugging is fairly similar to the flashing process. Ensuring the uniflash has been added to path. Similar to the OpenOCD method for debugging the process is separated into two processes. Including starting the debug server and calling the debug console.

After compiling, making sure the debug symbols have been included in the final output. The debug server can be started as such.

`make BOARD=cc2650stk debug-server`

This will start the debugging server. Running;

`make BOARD=cc2650stk debug`

Will start gdb with the appropriate initialisation for the target and debugging tool.

Like the process of flashing, the debugger can be changed with the 'XDEBUGGER' flag. Such as the following.

`make BOARD=cc2650stk debug-server XDEBUGGER=XDS100v3`

Which will use all the appropriate distribution files.

Let me know how you go.

Nicholas

nicholas-c-jackson commented:
@A-Paul How did you go?

A-Paul commented:
@nicholas-c-jackson, thanks for pinging.

I've finally found time to follow https://github.com/RIOT-OS/RIOT/pull/5419#issuecomment-219260570.  
But I got an error message when trying to flash.

> aar680@mobi31:~/git/RIOT/examples/hello-world$ make flash DEBUGGER=XDS110
> Building application "hello-world" for "cc2650stk" with MCU "cc26x0".
> 
> [...]
>    text    data     bss     dec     hex filename
>    7784     128    2872   10784    2a20 /home/aar680/git/RIOT/examples/hello-world/bin/cc2650stk hello-world.elf
> /opt/ti/uniflash_3.4/uniflash.sh -ccxml /home/aar680/git/RIOT/boards/cc2650stk/dist/cc26x0f128_XDS110.ccxml -program /home/aar680/git/RIOT/examples/hello-world/bin/cc2650stk/hello-world.elf

> 
> ***** Texas Instruments Universal Flash Programmer *****
> 
> <START: 19:00:36 GMT+0200 (MESZ)>
>
> \> Configuring the Flash Programmer with the given configuration ...
> 
> \> Flash Manager is configured for the following part: CC2650F128
> 
> \> Connecting to the target for Flash operations ... 
> 
> SEVERE: IcePick_C: Error connecting to the target: (Error -448 @ 0x0) Invalid spelling of the value of a configure variable. (Emulation package 6.0.14.5)

`BOARD=cc2650stk` is exported. The build also succeeded.
Any clues?

nicholas-c-jackson commented:
@A-Paul No problem. Are you testing with the XDS110 or XDS100v3?

To avoid conflicts with the the debugging flag, I've had to change it to XDEBUGGER, so by default it will choose the XDS110, if you're running the XDS100v3 that would explain the error.

If you're using the XDS110 with the DEVPACK, like this one, [http://www.ti.com/tool/cc-devpack-debug](http://www.ti.com/tool/cc-devpack-debug), from my experience they are notoriously flakey. But I've managed to figure out the best steps to ensure it works properly. 

First disconnect everything. Remove the DEVPACK from the STK, plug the usb into the DEVPACK, then plug the DEVPACK back onto the STK. If you plug the usb into the DEVPACK while still connected to the STK, it will be recognised as needing a firmware update, which is not true. This is a known issue from TI. I can seem to find the post in reference to it, if I find it I'll post it here.

If you're scenarios are none of the above I will continue to look into the problem.

nicholas-c-jackson commented:
@A-Paul Found what I was referring to, [http://processors.wiki.ti.com/index.php/Debug_DevPack_User_Guide#Device_Firmware_Upgrade](http://processors.wiki.ti.com/index.php/Debug_DevPack_User_Guide#Device_Firmware_Upgrade) Under the section Device Firmware Upgrade.

A-Paul commented:
Confirm I have a DevPack Rev:1.2.0. The kernel is showing a XDS110 when connected.

> [Fr Jun 17 16:18:52 2016] usb 2-1.4.1.3: New USB device found, idVendor=0451, idProduct=bef3
> [Fr Jun 17 16:18:52 2016] usb 2-1.4.1.3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [Fr Jun 17 16:18:52 2016] usb 2-1.4.1.3: Product: XDS110 (02.02.04.02) with CMSIS-DAP

The connection procedure comes close to shamanism and I'm proud to get involved in such secret knowledge. :D
But unfortunately I can't make progress with it.

A-Paul commented:
Additional Information:

- I get the **same** error code listed in https://github.com/RIOT-OS/RIOT/pull/5419#issuecomment-226550577, when just connecting *DevPack* without *cc2650stk*, and even when not connecting the *DevPack* at all.
- Flashing and running "hello-world" in master still works fine. Just to be sure that my setup isn't messed up.

nicholas-c-jackson commented:
@A-Paul I have not been able to replicate the issue on my machine.

I will try a clean VM image and see if the problem persists. 

Could you try navigating to the cc2650stk "dist" folder under the "boards" folder and with uniflash path set, run the following and paste your output.

> $UNIFLASH_PATH/ccs_base/common/uscif/dbgjtag -f cc26x0f128_XDS110.dat -rv -o -F inform,logfile=yes -S pathlength -S integrity

Thanks.

nicholas-c-jackson commented:
@A-Paul I set up a fresh VM with ubuntu 16.04 LTS and was able to flash and debug fine. So I'm not 100 percent sure what is going on.

 I found a thread on the TI forum that is remotely close to the issue you are having here [https://e2e.ti.com/support/development_tools/code_composer_studio/f/81/p/508785/1846111](https://e2e.ti.com/support/development_tools/code_composer_studio/f/81/p/508785/1846111). The only "solution" given was to reinstall Code Composer. So not really a solution. May want to try reflashing the firmware on the XDS110 with the following.

> xdsdfu -m
> and
> xdsdfu -f firmware.bin -r

In the $uniflash_path/ccs_base/common/uscif/xds110 folder.

Nicholas

A-Paul commented:
@nicholas-c-jackson, the output of *dbgjtag* is a lot of text. I pasted it in a [GIST](https://gist.github.com/A-Paul/6b821024994c8c47a8d0dbffcccfb0e6).

I also reflashed  the DevPack. No changes.

A-Paul commented:
When commenting out the property tag with `id="XDS110 Aux Port"` in *boards/cc2650stk/dist/cc26x0f128_XDS110.ccxml* flash is working again.

A-Paul commented:
Because of the flash issue I haven't tried to run the debug server. Here is what happens when connecting with `make debug`:

> aar680@mobi31:~/git/RIOT/examples/hello-world$ make debug-server             
> /opt/ti/uniflash_3.4/ccs_base/common/uscif/gdb_agent_console -p 3333 /home/aar680/git/RIOT/boards/cc2650stk/dist/cc26x0f128_XDS110.dat
> Successfully configured /home/aar680/git/RIOT/boards/cc2650stk/dist/cc26x0f128_XDS110.dat
> CPU Name             Port
> --------             ----
> cortex_m3_0         :3333
> 
> Starting all cores
> CPU Name             Status
> --------             ------
> cortex_m3_0          Waiting for client
> cortex_m3_0          Client connected...Connecting to Target
> FAILED! GTI_CONNECT(hpid=0x8c63348)=-1
> (Error -6306) Failed to connect to PRSC module.
> (Emulation package 6.0.14.5)
> 
> Connection failed...exiting

nicholas-c-jackson commented:
@A-Paul My apologies for going dark for too long, I had a fairly large paper to submit.

I have made some more commits including changing the JTAG modes from 2 pin cJTAG to 4 pin JTAG. From what I have taken from the material I have read, although cJTAG has not been implemented for the cc2650stk / launchxl, it should have defaulted to 4 pin JTAG I suspect a difference in our XDC tools / XDS Emulation or the various other components used.

I have tested both the XDS110 and XDS100v3 for flashing an debugging and are working on my end. If you could please test with the updated ccxml and dat files and let me know how you go. I'm still not 100% sure why I have the ability to flash and debug and you do not.

Nicholas

A-Paul commented:
Hi @nicholas-c-jackson. No problem, take your time. :-)
I've just made a quick test with *examples/hello_world*. `make flash` and `make term` is working now!
But now it turns out that `make reset` doesn't do anything. Resetting with `.../uniflash.sh -ccxml blah -reset` works.
Do you also feel responsible for that? ;-)

OlegHahm commented:
@mfrey, this might be interesting for you, too.

nicholas-c-jackson commented:
@A-Paul Sure why not ;)

I've added a new commit that implements the make flags for target resetting. I have tested it for both the XDS110 and the XDS100v3. It should be noted that the debugger will not respond to commands with the debug server running.

It's interesting that it works now. I assume there is differences between the emulation package from your machine and my machine. I assume they have changed how it handles cJTAG, now, instead of defaulting to JTAG it must just error out. I have looked everywhere and have not been able to find that particular error code. I need to make friends with the developers over at TI.

Nicholas

nicholas-c-jackson commented:
@A-Paul How have you gone with the latest changes?

miri64 commented:
Let's try to merge it for the next release.

A-Paul commented:
Hey @nicholas-c-jackson.
I checked the `make debug-server` `make debug` scenario (with some support from @cgundogan :-) and it works. There are issues in TI's gdbserver emulation, but this isn't this PRs fault.

> It should be noted that the debugger will not respond to commands with the debug server running.
This may be mentioned in the wiki. I confused *debugger* with *gdb* for a while. :D
Better refer to the debug interface as *XDS110* as *DevPack*
Beside thet calling `make reset` while being connected with `make term` works fine.

Thank you for addressing the comments and suggestions.

Please sqash your commits and rebase.

miri64 commented:
I take this as an ACK. ;-)

A-Paul commented:
> I take this as an ACK. ;-)

@miri64, that was my intention ;-)
Is an explicit "ACK" still recommended?

miri64 commented:
At least I'm working under the assumption. We really should discuss this on the maintainers list ^^

miri64 commented on boards/cc2650stk/Makefile.include:
```diff
@@ -11,4 +12,16 @@ include $(RIOTBOARD)/Makefile.include.serial
+export DEBUGSERVER_FLAGS = -p 3333 $(RIOTBOARD)/$(BOARD)/dist/$(CPU_MODEL)_$(XDEBUGGER).dat
+
+# configure the debugging tool
+export DEBUGGER = $(PREFIX)gdb 
```
Please remove trailing whitespace (as reported by Murdock)

miri64 commented on boards/cc2650stk/dist/cc26x0f128_XDS100v3.ccxml:
```diff
@@ -0,0 +1,17 @@
+    </configuration>
+</configurations>
+
+
```
Please remove additional newline (as reported by Murdock)

nicholas-c-jackson commented:
Sorry for the delay I've recently been on sick leave. 

I have removed the newline and whitespace from the aforementioned files. 

Slight issue, this is my first time contributing to an opensource project of this size. With personal and work repositories I've always had control over each branch so my knowledge is lacking in what to do next.

I have rebased the 9 commits into a single commit message but I am unsure in the following procedures a little advice would be great.

Nicholas

nicholas-c-jackson commented:
Squashed and rebased, I followed several articles, if you could please double check that would be great.

Nicholas

A-Paul commented:
Hi @nicholas-c-jackson 
> Sorry for the delay I've recently been on sick leave.

Don't worry. You're pretty fast. :-)

> I have rebased the 9 commits into a single commit message but I am unsure in the following procedures a little advice would be great.

You have done what is required at this point.
In general your work is acknowledged and will be merged into *master* when the CI-Tool runs without errors.
If anything has to be changed you'll get a message.
Just stay calm ;-)

Thanks for your efforts so far...

miri64 commented:
Error to dtls-echo is unrelated (see #6032). Go.

miri64 commented:
Thanks for your contribution @nicholas-c-jackson! :-)

nicholas-c-jackson commented:
Yay, thank you @A-Paul @mfrey and @miri64. Feels good.

I will be over the coming days updating the documentation with possibly writing about integrations of gdb with TI emulation tools through CCS.

shaolinkhoa commented:
Hi,

I just want to update about flashing cc2650
The current version of my uniflash is uniflash_4.1. 
I can't find `uniflash.sh` in UNIFLASH_PATH so I can't use `make BOARD=cc2650stk flash` 

To boot code.hex into cc2650 luanchpad, I have to launch Uniflash and load file hex manually.

Do you know how to fix it ?

yogo1212 commented:
Hi :-)

in my installation of uniflash 4.1 `uniflash.sh` is present in the top level of the installation directory (`$HOME/ti/uniflash`... by default).

if it is somehow missing in your installation, you should be able to just use this uniflash.sh:
```
#! /bin/sh
# This script executes UniFlash command line support

export CURPATH=$(dirname $(which $0))

export UNIFLASH_PATH=${CURPATH}/ccs_base/scripting/examples/uniflash/cmdLine/

export TI_TRACE_DISABLE=1

${CURPATH}/ccs_base/scripting/bin/dss.sh ${UNIFLASH_PATH}/uniFlash_main.js "$@"
```
it should sit inside the mentioned directory (next to `uniflashCLI.sh`)

shaolinkhoa commented:
hi @yogo1212 ,
Thank you for helping me. It shows that I don't have dss.sh:
```
/home/khoa/ti/uniflash_4.1/uniflash.sh -ccxml /home/khoa/RIOT/boards/cc2650stk/dist/cc26x0f128_XDS110.ccxml -program /home/khoa/RIOT/riot-apps/uart_shellcommands/bin/cc2650stk/uart_shellcommands.elf
/home/khoa/ti/uniflash_4.1/uniflash.sh: 10: /home/khoa/ti/uniflash_4.1/uniflash.sh: /home/khoa/ti/uniflash_4.1/ccs_base/scripting/bin/dss.sh: not found
/home/khoa/RIOT/riot-apps/uart_shellcommands/../../Makefile.include:342: recipe for target 'flash' failed
make: *** [flash] Error 127
```

p/s: Have you finished the RF example for CC2650 ? I would like to run RF of CC2650 on RIOT OS .
